pattern/variable/Variable.java

```diff
@@ -74,6 +81,34 @@ public ThingVariable asThing() {
         throw GraknException.of(INVALID_CASTING, className(this.getClass()), className(ThingVariable.class));
     }
 
+    public void addHints(Set<Label> labels) {
```

Comment:
I thought i had pointed this out before hmmm maybe I didn't hit submit:
Unless you can convince me otherwise, I think the API for adding/retaining/modifying type hints should _not_ be overlapping with the `isSatisfiable()` API internally. An external user can `setSatisfiable(false/true)` from outside, but from the variable's point of view, it is disconnected from type hints completely. This means the following methods should be simple adders/removers/setters, and the external user should use them to determine satisfiability (if using hints to do this at all!)

Reply:
There was a misalignment with the use of `isSatisfiable` here. I've now moved the logic of changing `isSatisfiable` to outside of `Variable`, which means that we have to do these checks when some of these methods are called. In particular, I've made those checks on `TypeHinter`, and also on `Rule` where the `TypeHints` are pruned based on the `TypeHints` on the when conjunction. I suggest you confer that's the correct behaviour.